trl/trainer/utils.py

```diff
@@ -1781,6 +1781,32 @@ def selective_log_softmax(logits, index) -> torch.Tensor:
     return per_token_logps
 
 
+def rowise_entropy(logits, chunk_size: int = 2048) -> torch.Tensor:
+    """
+    Computes the entropy of a tensor of shape [B, S, V] row-wise to reduce the memory
+    consumed by the softmax operation.
+
+    Args:
+        logits (`torch.Tensor`):
+            Logits tensor of shape `(..., num_classes)`.
+        chunk_size (`int`, *optional*, defaults to `2048`):
+            The number of chunks the batch will be split into to reduce memory consumption.
+    Returns:
+        `torch.Tensor`:
+            Entropy of each position in a sequence
+    """
+    per_token_entropies = []
+    for i in range(0, logits.shape[0], chunk_size):
+        logits_chunk = logits[i : i + chunk_size]
+        for row_logits in logits_chunk:  # loop to reduce peak mem consumption
+            row_logps = F.log_softmax(row_logits, dim=-1)
+            row_entropy = -(torch.exp(row_logps) * row_logps).sum(-1)
+            per_token_entropies.append(row_entropy)
```

Comment:
doesn't the inner loop negate the purpose of `chunk_size`? we can achieve the same thing by:
```suggestion
    for logits_chunk in logits.split(chunk_size, dim=0):
        logps = F.log_softmax(logits_chunk, dim=-1)    
        chunk_entropy = -(torch.exp(logps) * logps).sum(-1)
        per_token_entropies.append(chunk_entropy)
```

with a default `chunk_size=1`.